/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package stock;

import bean.CGenUtil;
import bean.ClassEtat;
import bean.ClassMere;
import bean.*;
import com.google.gson.Gson;
import fabrication.Fabrication;
import fabrication.FabricationFille;
import fabrication.OfFille;
import faturefournisseur.As_BonDeLivraison;
import faturefournisseur.FactureFournisseur;
import faturefournisseur.Fournisseur;
import ferme.couvoir.Eclosion;
import inventaire.Inventaire;
import inventaire.InventaireFille;
import inventaire.InventaireFilleCpl;
import java.sql.Connection;
import java.sql.Date;
import java.util.ArrayList;
import java.util.List;

import magasin.Magasin;
import maintenance.configuration.CompteurMaintenance;
import mg.cnaps.compta.ComptaEcriture;
import mg.cnaps.compta.ComptaSousEcriture;
import mg.cnaps.compta.ConstanteCompta;
import paramCompta.ParamComptaStatique;
import produits.Ingredients;
import utilitaire.ConstanteEtat;
import utilitaire.UtilDB;
import utilitaire.Utilitaire;
import utils.ConstanteSocobis;
import java.sql.Statement;
import utils.ConstanteStation;
import vente.As_BondeLivraisonClient;

public class MvtStock extends ClassMere {

    private String id, designation, idMagasin, idVente, idTransfert, idTypeMvStock,idPoint,idobjet,fabPrecedent,idCategorieStock ;
    private Date daty;

    EtatStock [] etatStock;
    EtatStockParEntree[] etatStockParEntree;
    String heure;

    public String getHeure() {
        return heure;
    }

    public void setHeure(String heure) {
        if((this.getMode().compareToIgnoreCase("modif")==0||this.getMode().compareToIgnoreCase("insert")==0)&&(heure==null||heure.equals(""))){
            heure=Utilitaire.heureCouranteHMS();
        }
        this.heure = heure;
    }

    @Override
    public boolean isSynchro(){
        return true;
    }
    
    public MvtStock() throws Exception {
        this.setNomTable("MVTSTOCK");
        this.setNomClasseFille("stock.MvtStockFille");
        this.setLiaisonFille("idMvtStock");
	 
    }

    public EtatStock[] getEtatStock() {
        return etatStock;
    }

    public void setEtatStock(EtatStock[] etatStock) {
        this.etatStock = etatStock;
    }

    public EtatStockParEntree[] getEtatStockParEntree() {
        return etatStockParEntree;
    }

    public void setEtatStockParEntree(EtatStockParEntree[] etatStockParEntree) {
        this.etatStockParEntree = etatStockParEntree;
    }

    public String getIdCategorieStock() {
        return idCategorieStock;
    }

    public void setIdCategorieStock(String idCategorieStock)  throws Exception{
        if ((this.getMode().compareToIgnoreCase("modif")==0) && Utilitaire.champNull(idCategorieStock).compareTo("") == 0 ) {
            throw new Exception("Champ cat\u00E9gorie de stock obligatoire");
        }
        this.idCategorieStock = idCategorieStock;
    }

    public String getFabPrecedent() {
        return fabPrecedent;
    }

    public void setFabPrecedent(String fabPrecedent) {
        this.fabPrecedent = fabPrecedent;
    }

    public String getId() {
        return id;
    }

    public void setId(String id) {
        this.id = id;
    }

    public String getDesignation() {
        return designation;
    }

    public void setDesignation(String designation) {
        this.designation = designation;
    }

    public String getIdMagasin() {
        return idMagasin;
    }

    public void setIdMagasin(String idMagasin) throws Exception{
        if ((this.getMode().compareToIgnoreCase("modif")==0||this.getMode().compareToIgnoreCase("insert")==0) && Utilitaire.champNull(idMagasin).compareTo("") == 0 ) {
            throw new Exception("Champ magasin obligatoire");
        }
        this.idMagasin = idMagasin;
    }

    public String getIdVente() {
        return idVente;
    }

    public void setIdVente(String idVente) {
        this.idVente = idVente;
    }

    public String getIdTransfert() {
        return idTransfert;
    }

    public void setIdTransfert(String idTransfert) {
        this.idTransfert = idTransfert;
    }

    public String getIdTypeMvStock() {
        return idTypeMvStock;
    }

    public void setIdTypeMvStock(String idTypeMvStock) throws Exception{
        if ((this.getMode().compareToIgnoreCase("modif")==0||this.getMode().compareToIgnoreCase("insert")==0) && (idTypeMvStock== null || idTypeMvStock.compareToIgnoreCase("") == 0)) {
            throw new Exception("Champ type du mouvement stock obligatoire");
        }
        this.idTypeMvStock = idTypeMvStock;
    }

    public Date getDaty() {
        return daty;
    }

    public void setDaty(Date daty) {
        if((this.getMode().compareToIgnoreCase("modif")==0)&&(daty==null||daty.equals(""))){
            daty=Utilitaire.dateDuJourSql();
        }
        this.daty = daty;
    }

    public String getIdPoint() {
        return idPoint;
    }

    public void setIdPoint(String idPoint) throws Exception {
        if ((this.getMode().compareToIgnoreCase("modif")==0||this.getMode().compareToIgnoreCase("insert")==0) && (idPoint== null || idPoint.compareToIgnoreCase("") == 0)) {
            throw new Exception("Champ point obligatoire");
        }
        this.idPoint = idPoint;
    }

    public String getIdobjet() {
        return idobjet;
    }

    public void setIdobjet(String idobjet) {
        this.idobjet = idobjet;
    }
    

    @Override
    public String getTuppleID() {
        return id;
    }

    @Override
    public String getAttributIDName() {
        return "id";
    }

    @Override
    public void construirePK(Connection c) throws Exception {
        this.preparePk("MVTST", "GETSEQMVTSTOCK");
        this.setId(makePK(c));
    }

    protected void controlerMvt(Connection c) throws Exception {
        if ((this.getMode().compareToIgnoreCase("modif")==0||this.getMode().compareToIgnoreCase("insert")==0) && (this.getIdMagasin()== null || this.getIdMagasin().compareToIgnoreCase("") == 0)) {
            throw new Exception("Champ magasin obligatoire");
        }
    }

    public Magasin getMagasin(Connection c) throws Exception {
        if (c == null) {
            throw new Exception("Connection non etablie");
        }
        Magasin magasin = new Magasin();
        magasin.setId(this.getIdMagasin());
        Magasin[] magasins = (Magasin[]) CGenUtil.rechercher(magasin, null, null, c, " ");
        if (magasins.length > 0) {
            return magasins[0];
        }
        return null;
    }

    public MvtStockFille[] getMvtStockFilles(Connection c) throws Exception {
        boolean estOuvert = false;
        try {
            if (c == null) {
                estOuvert = true;
                c = new UtilDB().GetConn();
            }
            MvtStockFille msf = new MvtStockFille();
            msf.setIdMvtStock(this.getId());
            MvtStockFille[] msfs = (MvtStockFille[]) CGenUtil.rechercher(msf, null, null, c, " ");
            if (msfs.length > 0) {
                return msfs;
            }
            return null;
        } catch (Exception e) {
            throw e;
        } finally {
            if (c != null && estOuvert == true) {
                c.close();
            }
        }
    }

    public void createInventaireZero(String u, Connection c) throws Exception {
    if (this.getIdTypeMvStock() == ConstanteStation.TYPEMVTSTOCKENTREE) {
        MvtStockFille[] msfs = getMvtStockFilles(c);
        this.setFille(msfs);
        if (msfs != null && msfs.length > 0) {
            for (MvtStockFille mvf : msfs) {
                InventaireFilleCpl invFCpl = new InventaireFilleCpl();
                invFCpl.setIdMagasin(this.getIdMagasin());
                invFCpl.setIdProduit(mvf.getIdProduit());
                InventaireFilleCpl[] invFCpls = invFCpl.getInventaireFilles(c);
                if (invFCpls == null) {
                    Magasin m = getMagasin(c);
                    Inventaire inv = (Inventaire) m.generateInventaireMere().createObject(u, c);
                    InventaireFille invF = inv.generateInventaireFilleZero();
                    invF.setIdProduit(mvf.getIdProduit());
                    invF.createObject(u, c);
                    inv.validerObject(u, c);
                }
            }
        }
    }
}
    public OfFille[] getOfFilles(String nt,Connection c) throws Exception {
        String nomTable="fabricationOfFille";
        if(nt!=null&&nt.compareToIgnoreCase("")!=0) nomTable=nt;
        OfFille crt=new OfFille();
        crt.setNomTable(nomTable);
        crt.setIdFab(this.getIdobjet());
        return (OfFille[]) CGenUtil.rechercher(crt, null, null, c, "");
    }
    public FabricationFille[] getFabricationFilles(String nt,Connection c) throws Exception {
            String nomTable="FABRICATIONFILLE";
            if(nt!=null&&nt.compareToIgnoreCase("")!=0) nomTable=nt;
            FabricationFille crt=new FabricationFille();
            crt.setNomTable(nomTable);
            crt.setIdMere(this.getIdobjet());
            return (FabricationFille[]) CGenUtil.rechercher(crt, null, null, c, "");
    }
    public MvtStockFille[] getMvtStockPrecedent(String nt,Connection c) throws Exception {
           String nomTable="MVTSTOCKFille";
           if(nt!=null&&nt.compareToIgnoreCase("")!=0) nomTable=nt;
           MvtStockFille crt=new MvtStockFille();
           return null;
    }
    public static MvtStockFille[] getMvtStockEntree(MvtStockFille[] msfs) throws Exception {
            ArrayList<MvtStockFille> msfList = new ArrayList<MvtStockFille>();
            for (MvtStockFille msf : msfs) {
                if(msf.getEntree()>0) msfList.add(msf);
            }
            return msfList.toArray(new MvtStockFille[msfList.size()]);
    }
    public static MvtStockFille getMvtStockEntreeCorrespondantFabFille(MvtStockFille[] msfs,FabricationFille[] listeFab) throws Exception {
        MvtStockFille[] mvEntree=getMvtStockEntree(msfs);
        for (MvtStockFille msf : mvEntree) {
            for (FabricationFille fab : listeFab) {
                if(msf.getIdProduit().compareToIgnoreCase(fab.getIdIngredients())==0) return msf;
            }
        }
        return null;
    }

    /*@Override
    public Object validerObject(String u, Connection c) throws Exception {
        MvtStockFille[]fille=this.getMvtStockFilles(c);
        if (this.getIdTypeMvStock()!=null && this.getIdTypeMvStock().compareToIgnoreCase(ConstanteSocobis.TYPE_MVT_SORTIE)==0) {
            checkResteStock(c,fille, this);
            createGenerateMvt(u,c,fille);
        }
        for (int i = 0; i < fille.length; i++) {
            fille[i].updateQteIngredient(u, c);
        }
        super.validerObject(u, c);
        // createInventaireZero(u, c);
        if(this.getFabPrecedent()==null||this.getFabPrecedent().compareToIgnoreCase("")==0) return this;
        Fabrication precedent=new Fabrication();
        precedent.setId(this.getFabPrecedent());
        FabricationFille[] ofFille=(FabricationFille[])precedent.getFille(null, c,"");
        if(ofFille==null||ofFille==null) return this;
        Fabrication fab=new Fabrication();
        fab.setId(this.getFabPrecedent());
        //MvtStockFille[] mvtFilleFab=fab.getMvtStockFille(null,c);
        //if(MvtStock.getMvtStockEntreeCorrespondantFabFille(mvtFilleFab,ofFille)!=null) return this;
        for(int i=0;i<fille.length;i++){
            if(fille[i].getSortie()>0)
            {
                if(fille[i].getIdProduit().compareToIgnoreCase(ofFille[0].getIdIngredients())==0)
                {
                    MvtStock mvt=(MvtStock) this.dupliquerSansBase();
                    mvt.construirePK(c);
                    mvt.setIdobjet(this.getFabPrecedent());
                    mvt.setFabPrecedent(null);
                    mvt.setIdTypeMvStock(ConstanteStation.TYPEMVTSTOCKENTREE);
                    MvtStockFille nouveau=fille[i].contrer();
                    nouveau.setIdMvtStock(mvt.getId());
                    nouveau.construirePK(c);
                    MvtStockFille[]ln={nouveau};
                    mvt.setFille(ln);
                    mvt.createObject(u, c);
                    mvt.validerObject(u, c);
                }
            }
        }
        return this;
    }*/


    EtatStock[] findEtatStockGlobale(Connection c) throws Exception
    {
        if(this.getEtatStock()!=null)return this.getEtatStock();
        EtatStock crt=new EtatStock();
        crt.setIdMagasin(this.getIdMagasin());
        MvtStockFille[] invFille=(MvtStockFille[]) this.getFille();
        String listeWhere=" and idProduit in ("+ Utilitaire.tabToString(invFille,"idProduit","'",",")+")";
        EtatStock[] retour= (EtatStock[]) CGenUtil.rechercherReq(new EtatStock(),EtatStock.getReqEtatStock(Utilitaire.dateDuJour()),c,listeWhere);
        this.setEtatStock(retour);
        return retour;
    }

    EtatStockParEntree[] findEtatStockEntree(Connection c) throws Exception{
        if(this.getEtatStockParEntree()!=null) return this.getEtatStockParEntree();
        EtatStockParEntreeStandard crt=new EtatStockParEntreeStandard();
        crt.setIdMagasin(this.getIdMagasin());
        MvtStockFille[] invFille=(MvtStockFille[]) this.getFille();
        String listeWhere=" and idProduit in ("+ Utilitaire.tabToString(invFille,"idProduit","'",",")+")";
        EtatStockParEntree[] retour= (EtatStockParEntree[]) CGenUtil.rechercherReq(crt,EtatStockParEntreeStandard.getRequeteEtatStockEntree(Utilitaire.dateDuJour()),c,listeWhere);
        this.setEtatStockParEntree(retour);
        return retour;
    }


    MvtStock contrer(Connection c) throws Exception
    {
        MvtStock retour=(MvtStock) this.dupliquerSansBase();
        MvtStockFille[] lFille=this.getMvtStockFilles(c);
        this.getMvtStockFilles(c);
        MvtStockFille[] filleAret=new MvtStockFille[lFille.length];
        retour.construirePK(c);
        for(int i=0;i<lFille.length;i++)
        {
            filleAret[i]=lFille[i].contrer();
            filleAret[i].construirePK(c);
            filleAret[i].setIdMvtStock(retour.getId());
        }
        retour.setFille(filleAret);
        if(retour.getIdTypeMvStock()!=null&&retour.getIdTypeMvStock().compareToIgnoreCase(ConstanteSocobis.TYPE_MVT_SORTIE)==0)
        {
            retour.setIdTypeMvStock(ConstanteSocobis.TYPE_MVT_ENTREE);
        }
        else if(retour.getIdTypeMvStock()!=null&&retour.getIdTypeMvStock().compareToIgnoreCase(ConstanteSocobis.TYPE_MVT_ENTREE)==0) retour.setIdTypeMvStock(ConstanteSocobis.TYPE_MVT_SORTIE);
        return retour;
    }



    public void releverGaz(String u, Connection c, String idLigne, MvtStockFille fille) throws Exception {

        CompteurMaintenance ancien = new CompteurMaintenance().getLastCompteur(c, idLigne, fille.getIdProduit());
        CompteurMaintenance cm = new CompteurMaintenance();
        cm.setDaty(this.getDaty());
        cm.setIdMagasin(this.getIdMagasin());
        cm.setIdCategorie(fille.getIdProduit());
        cm.setIdLigne(idLigne);
        if (ancien != null) cm.setAncien(ancien.getValeur());
        else cm.setAncien(0);
        double entreeKilogram = fille.getEntree(); // En lilogram
        double entreeLitre = entreeKilogram/0.577; // En litre
        double valuerPourcentage = (entreeLitre/ConstanteSocobis.MAX_GAZ_VALEUR)*100;
        cm.setValeur(valuerPourcentage+cm.getAncien());
        CompteurMaintenance created = (CompteurMaintenance) cm.createObject(u,c);
        created.validerObject(u, c);


    }

    @Override
    public Object validerObject(String u, Connection c) throws Exception {
        Statement st =null;
        try {
            MvtStockFille[] fille = (MvtStockFille[]) this.getFille("mvtStockFilleCptStock",c,"");

            //this.genererEcriture(u,c,fille);

            if (fille == null) fille = this.getMvtStockFilles(c);
            setFille(fille);
            if (this.getIdTypeMvStock() != null && this.getIdTypeMvStock().compareToIgnoreCase(ConstanteSocobis.TYPE_MVT_SORTIE) == 0) {
                checkResteStock(c, fille, this);
                //createGenerateMvt(u, c, fille);
            }
            st=c.createStatement();
            for (int i = 0; i < fille.length; i++) {
                fille[i].updateQteIngredientBatch(u, st);
            }
            st.executeBatch();

            Magasin magasin = new Magasin();
            magasin.setNomTable("magasin2");
            magasin.setId(this.getIdMagasin());

            Magasin[] magasins = (Magasin[]) CGenUtil.rechercher(magasin, null, null, c, "");

            for (MvtStockFille msf : fille)
            {
                if (msf.getIdProduit().equalsIgnoreCase(ConstanteSocobis.gaz))
                {
                    this.releverGaz(u, c, magasins[0].getIdLigne(), msf);
                }
            }


            this.setHeure(Utilitaire.heureCouranteHMS());
            this.setDaty(Utilitaire.dateDuJourSql());
            super.validerObject(u, c);
            this.setFille(fille);
            // creation inventaire vide lors de chaque saisie entree de stock
            /*
            if (this.getIdTypeMvStock().equalsIgnoreCase(ConstanteStation.TYPEMVTSTOCKENTREE)) {
                Inventaire inv = new Inventaire();
                inv.preparePk("IVTVIDE", "GETSEQINVENTAIRE");
                inv.setId(inv.makePK());
                inv.setDaty(this.getDaty());
                inv.setDesignation("Inventaire auto vide du mouvement de stock num "+this.getId());
                inv.setIdMagasin(this.getIdMagasin());
                inv.setRemarque("Inventaire auto vide");
                InventaireFille[] invfille = new InventaireFille[getFille().length];
                for (int i = 0; i < this.getFille().length; i++) {
                    invfille[i] = new InventaireFille();
                    invfille[i].setIdInventaire(inv.getId());
                    invfille[i].setDaty(this.getDaty());
                    invfille[i].setIdProduit(((MvtStockFille)this.getFille()[i]).getIdProduit());
                    invfille[i].setExplication(inv.getRemarque());
                    invfille[i].setQuantiteTheorique(0);
                    invfille[i].setQuantite(0);
                    invfille[i].setMvtsrc(((MvtStockFille)this.getFille()[i]).getMvtSrc());
                    invfille[i].setPu(((MvtStockFille)this.getFille()[i]).getPu());
                }
                inv.setFille(invfille);
                inv.createObject(u, c);
                inv.validerObject(u, c);
            }*/
            String idObjet = getIdobjet();
            if (idObjet != null && idObjet.startsWith("BLC")) {
                processUpdateStockEngage(u, c);
            }

            // createInventaireZero(u, c);
            if (this.getFabPrecedent() == null || this.getFabPrecedent().compareToIgnoreCase("") == 0) return this;
            Fabrication precedent = new Fabrication();
            precedent.setId(this.getFabPrecedent());
            FabricationFille[] ofFille = (FabricationFille[]) precedent.getFille(null, c, "");
            if (ofFille == null || ofFille == null) return this;
            Fabrication fab = new Fabrication();
            fab.setId(this.getFabPrecedent());
            //MvtStockFille[] mvtFilleFab=fab.getMvtStockFille(null,c);
            //if(MvtStock.getMvtStockEntreeCorrespondantFabFille(mvtFilleFab,ofFille)!=null) return this;
            for (int i = 0; i < fille.length; i++) {
                if (fille[i].getSortie() > 0) {
                    if (fille[i].getIdProduit().compareToIgnoreCase(ofFille[0].getIdIngredients()) == 0) {
                        MvtStock mvt = (MvtStock) this.dupliquerSansBase();
                        mvt.construirePK(c);
                        mvt.setIdobjet(this.getFabPrecedent());
                        mvt.setFabPrecedent(null);
                        mvt.setIdTypeMvStock(ConstanteStation.TYPEMVTSTOCKENTREE);
                        MvtStockFille nouveau = fille[i].contrer();
                        nouveau.setIdMvtStock(mvt.getId());
                        nouveau.construirePK(c);
                        MvtStockFille[] ln = {nouveau};
                        mvt.setFille(ln);
                        mvt.createObject(u, c);
                        mvt.validerObject(u, c);
                    }
                }
            }
            return this;
        }
        catch (Exception e)
        {
            e.printStackTrace();
            throw e;
        }
        finally {
            if(st!=null) st.close();
        }
    }

    public As_BondeLivraisonClient getBL(Connection c) throws Exception {
        String idObjet = getIdobjet();
        if (idObjet != null && idObjet.startsWith("BLC")) {
            return (As_BondeLivraisonClient) new As_BondeLivraisonClient().getById(idObjet, null, c);
        }

        return null;
    }

    public void processUpdateStockEngage(String u, Connection c) throws Exception {
        As_BondeLivraisonClient blAssocie = getBL(c);
        if (blAssocie == null) {
            throw new Exception("Bon de livraison associé intouvable");
        }

        String idVente = blAssocie.getIdvente();

        MvtStock mvtSearch = new MvtStock();
        mvtSearch.setIdVente(idVente);
        mvtSearch.setNomTable("stock_engage");

        MvtStock[] resultatRecherche = (MvtStock[]) CGenUtil.rechercher(mvtSearch, null, null, c, "");
        if (resultatRecherche.length == 0) {
            throw new Exception("Stock avec engagement introuvable pour le vente : " + idVente + " associé a la blc : " + getIdobjet());
        }

        MvtStock stockEngage = resultatRecherche[0];
        MvtStockFille[] stockEngageFilles = (MvtStockFille[]) stockEngage.getFille("STOCK_ENGAGE_FILLE", c, "");

        MvtStockFille[] stockFille = (MvtStockFille[]) getFille();
        for (MvtStockFille fille : stockFille) {
            boolean finded = false;
            for (MvtStockFille filleEngage : stockEngageFilles) {
                if (fille.getIdProduit().equals(filleEngage.getIdProduit())) {
                    filleEngage.setSortie(filleEngage.getSortie() - fille.getSortie());
                    if (filleEngage.getSortie() < 0) {
                        throw new Exception("Quantité superieure à quantité engagé");
                    }
                    filleEngage.setNomTable("STOCK_ENGAGE_FILLE");
                    filleEngage.updateToTable(c);
                    finded = true;
                    break;
                }
            }

            if (!finded) {
                throw new Exception("Stock avec engagement introuvable pour le produit : " + fille.getIdProduit());
            }
        }
    }

    public void saveMvtStockFille(String u, Connection c) throws Exception {
        MvtStockFille[] mvtf = (MvtStockFille[]) this.getFille();
        for (int i = 0; i < mvtf.length; i++) {
            mvtf[i].setId(null);
            mvtf[i].setIdMvtStock(this.getId());
            mvtf[i].createObject(u, c);
        }
    }

    @Override
    public void controler(Connection c) throws Exception {
        super.controler(c);
        this.controlerMvt(c);
    }

    public void checkResteStock(Connection c, MvtStockFille[] fille, MvtStock mere) throws Exception {
        for (MvtStockFille mvtFille : fille) {
            if (mvtFille.estSuffisant(c, mere)==false) {
                throw new Exception(String.format("Quantit\u00E9 en stock insuffisante pour le produit %s dans le magasin", mvtFille.getIdProduit()));
            }
        }
    }

    
    public void createGenerateMvt(String refuser, Connection c, MvtStockFille [] mvf) throws Exception{
        List<MvtStockFille[]> mvtstockfillenew= new ArrayList<MvtStockFille[]>();
        for (int i = 0; i < mvf.length; i++) {
            String apres = "";
            apres += "AND IDPRODUIT = '"+mvf[i].getIdProduit()+"' AND IDMAGASIN='"+this.getIdMagasin()+"'";
            Ingredients ing = mvf[i].getIngredient(c);
            if (ing.getTypeStock().compareToIgnoreCase("FIFO")==0) {
                apres+=" order by daty asc,id asc "; 
            }else if (ing.getTypeStock().compareToIgnoreCase("LIFO")==0){
                apres += " order by daty desc,id desc ";
            } else{
                return;
            }
            MvtStockEntreeAvecReste[] liste=this.getMvtSelonType(c, apres);
            mvtstockfillenew.add(mvf[i].genererNouveauMvt(liste));
        }
        this.deleteObjectAncien(c,mvf);
        for (MvtStockFille[] mvtStockFilles : mvtstockfillenew) {
            this.createNewSortie(refuser, mvtStockFilles, c);
        }
    }

    public MvtStockEntreeAvecReste[] getMvtSelonType(Connection c,String apres) throws Exception {
        MvtStockEntreeAvecReste search = new MvtStockEntreeAvecReste();
        search.setIdMagasin(this.getIdMagasin());
        return  (MvtStockEntreeAvecReste[]) CGenUtil.rechercher(search,null, null, apres);
    }
    
    public void deleteObjectAncien(Connection c,MvtStockFille [] mvt) throws Exception {
        for (MvtStockFille fille : mvt) {
            fille.deleteToTable(c);
        }
    }

    public void createNewSortie(String refUser, MvtStockFille[] listeMvtFille, Connection c) throws Exception {
        for (MvtStockFille fille : listeMvtFille) {
            fille.createObject(refUser,c);
        }
    }


    public MvtStockFilleGrp[]  getDetailGrp(Connection c)  throws Exception
    {
        MvtStockFilleGrp mg = new MvtStockFilleGrp();
        mg.setNomTable("mvtStockFilleCptStock");
        MvtStockFilleGrp[] grp = (MvtStockFilleGrp[]) CGenUtil.rechercher(mg,null, null, c , " And IDMVTSTOCK = '"+ this.getId() +"'");
        if (grp.length == 0) {
            throw new Exception("Stock details introuvable pour l'ID : " + id );
        }
        return grp;
    }

    public As_BonDeLivraison getBLFournisseur(Connection c) throws Exception {
        String idObjet = getIdobjet();
        if (idObjet != null) {
            return (As_BonDeLivraison) new As_BonDeLivraison().getById(idObjet, null, c);
        }

        return null;
    }

    public Fournisseur getFournisseur(Connection c) throws Exception {
        String idObjet = getIdobjet();
        if (idObjet != null) {
            return (Fournisseur) new Fournisseur().getById(getBLFournisseur(c).getIdFournisseur(), null, c);
        }
        return null;
    }

    public FactureFournisseur  getFactureFournisseur(Connection c) throws Exception {
        String idObjet = getIdobjet();
        if (idObjet != null) {
            As_BonDeLivraison a = getBLFournisseur(c);
            FactureFournisseur[] b = (FactureFournisseur[]) CGenUtil.rechercher(new FactureFournisseur(), null, null,c," and idbc = '"+a.getIdbc()+"'");
            if(b.length > 0) {
                return b[0];
            }
        }
        return null;
    }
    public TransfertStock getTransfertStock(Connection c) throws Exception {
        TransfertStock transfertStock = new TransfertStock();
        transfertStock.setId(this.getIdTransfert());
        return (TransfertStock) CGenUtil.getMereAvecFille(transfertStock, c, "TRANSFERTSTOCKDETAILS");
    }

    public List<Double> comparerTransfertMvtStock( MvtStockFille[] grp,TransfertStock transfertStock) throws Exception {
        TransfertStockDetails[] transfertStockDetails = (TransfertStockDetails[]) transfertStock.getFille();
        List<Double> retour=new ArrayList<>();
        //double[] ecart = new double[transfertStockDetails.length] ;
        for (int i = 0; i < grp.length; i++) {
            TransfertStockDetails[] transfertStockDetails1 = (TransfertStockDetails[]) AdminGen.find(transfertStockDetails, new String[]{"id"}, new String[]{grp[i].getIdTransfertDetail()});
            double ecart = (double) (grp[i].getEntree() - transfertStockDetails1[i].getQuantite());
            if(ecart!=0) retour.add(ecart);
        }
        return retour;
    }

    public ComptaSousEcriture[] genererSousEcriture(Connection c,MvtStockFille[] mvtFilles) throws Exception {
        ComptaSousEcriture[] compta = null;
        if(mvtFilles==null||mvtFilles.length==0) return null;
        boolean canClose = false;
        if(this.getIdTransfert()!=null&&this.getIdTransfert().equalsIgnoreCase("")==false&&this.getFille()!=null&&((MvtStockFille)this.getFille()[0]).getSortie()>0)
            return null;

        try {
            if (c == null) {
                c = new UtilDB().GetConn();
                canClose = true;
            }

            MvtStockFilleGrp[] grp = this.getDetailGrp(c);
            compta = new ComptaSousEcriture[grp.length + mvtFilles.length];

            double montantEntreeTotal = 0;
            //double montantTvaTotal = getFactureFournisseur(c).getTVATotal(c);
            TransfertStock transfertStock = null;
            List<Double> ecart = null;
            if(this.getIdTransfert() != null&&this.getIdTransfert().compareToIgnoreCase("")!=0) {
                transfertStock = this.getTransfertStock(c);
                ecart = comparerTransfertMvtStock(grp, transfertStock);
            }
            if(ecart!=null&&ecart.size()>0&&this.getIdTransfert()!=null)
            {
                compta=new ComptaSousEcriture[ecart.size()*2];
                int j=0;
                for (; j < ecart.size(); j++) {
                    compta[j] = new ComptaSousEcriture();
                    compta[j].setRemarque(this.getId());
                    if (this.getIdTransfert() != null && grp[j].getEntree()>0) {
                        if (ecart.get(j) > 0) {
                            TypeObjet paramCompta = ParamComptaStatique.getById(ConstanteSocobis.gainStock);
                            compta[j].setLibellePiece(paramCompta.getVal());
                            compta[j].setCompte(paramCompta.getDesce());
                            compta[j].setCredit(ecart.get(j).doubleValue() *grp[j].getPu());
                        }
                        if (ecart.get(j) < 0) {
                            TypeObjet paramCompta = ParamComptaStatique.getById(ConstanteSocobis.perteStock);
                            compta[j].setLibellePiece(paramCompta.getVal());
                            compta[j].setCompte(paramCompta.getDesce());
                            compta[j].setDebit(ecart.get(j).doubleValue()*grp[j].getPu());
                        }
                    }
                }
                for(int indCredi=j;indCredi<compta.length;indCredi++) {
                    compta[indCredi]=new ComptaSousEcriture();
                    compta[indCredi].setLibellePiece("Compte de sttock du mouvement "+this.getId());
                    compta[indCredi].setRemarque(this.getId());
                    compta[indCredi].setCompte(mvtFilles[indCredi- ecart.size()].getCompte_stock());
                    if(ecart.get(indCredi- ecart.size()) > 0) compta[indCredi].setDebit(ecart.get(indCredi- ecart.size()).doubleValue()*grp[indCredi- ecart.size()].getPu());
                    else compta[indCredi].setCredit(ecart.get(indCredi- ecart.size()).doubleValue()*grp[indCredi- ecart.size()].getPu());
                }
                return compta;
            }

            int i = 0;
            for (; i < grp.length; i++) {
                MvtStockFilleGrp mvtStockFilleGrp = grp[i];
                double montantEntree = mvtStockFilleGrp.getMontantCalc();

                compta[i] = new ComptaSousEcriture();
                compta[i].setRemarque(this.getId());

                if(this.getIdTypeMvStock().compareToIgnoreCase(ConstanteSocobis.stockTypeInventaire)==0)  //Cas de type inventaire
                {
                    if(mvtStockFilleGrp.getEntree()>0) //Cas gain inventaire
                    {
                        compta[i].setLibellePiece("gain de Stock du mouvement " + this.getId()+" de l inventaire "+this.getIdobjet());
                        compta[i].setCompte(ConstanteSocobis.compteGainInvenatre);
                    }
                    else //Cas perte inventaire
                    {
                        compta[i].setLibellePiece("Perte de Stock du mouvement " + this.getId()+" de l inventaire "+this.getIdobjet());
                        compta[i].setCompte(ConstanteSocobis.comptePerteInvenatre);
                    }
                    //compta[i].setCompte(mvtStockFilleGrp.getCompte_sortie_stock());
                }
                else
                {
                    compta[i].setLibellePiece("Variation de Stock du mouvement " + this.getId());
                    compta[i].setCompte(mvtStockFilleGrp.getCompte_sortie_stock());
                }
                if(mvtStockFilleGrp.getEntree()>0) compta[i].setCredit(montantEntree);
                else compta[i].setDebit(montantEntree);
            }
            for(int indCredi=i;indCredi<compta.length;indCredi++) {
                compta[indCredi]=new ComptaSousEcriture();
                compta[indCredi].setLibellePiece("Compte de sttock du mouvement "+this.getId());
                compta[indCredi].setRemarque(this.getId());
                compta[indCredi].setCompte(mvtFilles[indCredi- grp.length].getCompte_stock());
                if(mvtFilles[indCredi- grp.length].getEntree()>0) compta[indCredi].setDebit(mvtFilles[indCredi-grp.length].getMontantCalc());
                else compta[indCredi].setCredit(mvtFilles[indCredi-grp.length].getMontantCalc());
            }

            // Ligne pour la TVA (si applicable)
            /*compta[i] = new ComptaSousEcriture();
            compta[i].setLibellePiece("TVA entrée Stock");
            compta[i].setRemarque("TVA entrée Stock");
            compta[i].setCompte(ConstanteStation.compteTVACollecte);
            compta[i].setDebit(montantTvaTotal);
            i++;*/

            // Ligne récapitulative fournisseur
            /*compta[i] = new ComptaSousEcriture();
            compta[i].setLibellePiece("Entrée Stock Total");
            compta[i].setRemarque("Entrée Stock Total");
            compta[i].setCompte(getFournisseur(c).getCompte());*/
            //compta[i].setCredit(montantEntreeTotal + montantTvaTotal);

        } catch (Exception e) {
            throw e;
        } finally {
            if (canClose && c != null) {
                c.close();
            }
        }
        return compta;
    }



    public void genererEcriture(String u, Connection c,MvtStockFille[]mvtFille) throws Exception{
        ComptaEcriture mere = new ComptaEcriture();
        Date dateDuJour = utilitaire.Utilitaire.dateDuJourSql();
        int exercice = utilitaire.Utilitaire.getAnnee(daty);
        mere.setDaty(dateDuJour);
        mere.setDesignation(this.getDesignation());
        mere.setExercice(""+exercice);
        mere.setDateComptable(this.getDaty());
        mere.setJournal(ConstanteCompta.journalOD);
        mere.setOrigine(this.getId());
        mere.setIdobjet(this.getId());

        ComptaSousEcriture[] filles = this.genererSousEcriture(c,mvtFille);
        if(filles==null) return;
        for(int i=0; i<filles.length; i++){
            filles[i].setIdMere(mere.getId());
            filles[i].setExercice(exercice);
            filles[i].setDaty(this.getDaty());
            filles[i].setJournal(ConstanteCompta.journalOD);

            if(filles[i].getDebit()>0 || filles[i].getCredit()>0) mere.ajouterFille(filles[i]);
        }
        mere.createObject(u, c);
    }

    public ComptaSousEcriture[] genererSousEcritureSortie(Connection c) throws Exception {
        ComptaSousEcriture[] compta = {};
        boolean canClose = false;
        try {
            if (c == null) {
                c = new UtilDB().GetConn();
                canClose = true;
            }

            MvtStockFilleGrp[] grp = this.getDetailGrp(c);
            compta = new ComptaSousEcriture[grp.length * 2];

            double montantSortieTotal = 0;

            int i = 0;
            for (int j = 0; j < grp.length; j++) {
                MvtStockFilleGrp mvtStockFilleGrp = grp[j];
                double montantEntree = mvtStockFilleGrp.getSortie();

                compta[i] = new ComptaSousEcriture();
                compta[i].setLibellePiece("Sortie Stock du compte - " + mvtStockFilleGrp.getCompte_sortie_stock());
                compta[i].setRemarque(mvtStockFilleGrp.getCompte_sortie_stock());
                compta[i].setCompte(mvtStockFilleGrp.getCompte_sortie_stock());
                compta[i].setDebit(montantEntree);

                i ++;
                compta[i] = new ComptaSousEcriture();
                compta[i].setLibellePiece("Stock du compte - " + mvtStockFilleGrp.getCompte_stock());
                compta[i].setRemarque(mvtStockFilleGrp.getCompte_stock());
                compta[i].setCompte(mvtStockFilleGrp.getCompte_stock());
                compta[i].setCredit(montantEntree);

                montantSortieTotal += montantEntree;
                i++;
            }

            // Ligne de la vente
//            compta[i] = new ComptaSousEcriture();
//            compta[i].setLibellePiece("Sortie Stock Total");
//            compta[i].setRemarque("Sortie Stock Total");
//            compta[i].setCompte(ConstanteCompta.variationDeStock);
//            compta[i].setDebit(montantSortieTotal);

        } catch (Exception e) {
            throw e;
        } finally {
            if (canClose && c != null) {
                c.close();
            }
        }
        return compta;
    }

    public void genererEcritureSortie(String u, Connection c) throws Exception {
        ComptaEcriture mere = new ComptaEcriture();
        Date dateDuJour = utilitaire.Utilitaire.dateDuJourSql();
        int exercice = utilitaire.Utilitaire.getAnnee(daty);
        mere.setDaty(dateDuJour);
        mere.setDesignation(this.getDesignation());
        mere.setExercice("" + exercice);
        mere.setDateComptable(this.getDaty());
        mere.setJournal(ConstanteCompta.journalOD);
        mere.setOrigine(this.getId());
        mere.setIdobjet(this.getId());
        mere.createObject(u, c);

        ComptaSousEcriture[] filles = this.genererSousEcritureSortie(c);
        for (int i = 0; i < filles.length; i++) {
            filles[i].setIdMere(mere.getId());
            filles[i].setExercice(exercice);
            filles[i].setDaty(this.getDaty());
            filles[i].setJournal(ConstanteCompta.journalOD);

            if (filles[i].getDebit() > 0 || filles[i].getCredit() > 0) filles[i].createObject(u, c);
        }
    }
    public void jeter(Connection c , String u) throws Exception {
        boolean canClose = false;
        try {
            if (c == null) {
                c = new UtilDB().GetConn();
                c.setAutoCommit(false);
                canClose = true;
            }
            this.setFille(this.getFille("mvtstockfillelib", c, ""));
            MvtStock dechetMere = (MvtStock) this.dupliquerSansBase();
            dechetMere.setId(null);
            dechetMere.setIdobjet(this.getId());
            dechetMere.setIdTypeMvStock(ConstanteSocobis.TYPE_MVT_SORTIE);
            MvtStockFille[] filles = (MvtStockFille[]) dechetMere.getFille();
            for (int i = 0; i < filles.length; i++) {
                filles[i].setId(null);
                filles[i].setSortie(filles[i].getEntree());
                filles[i].setEntree(0);
            }
            dechetMere.setFille(filles);
            dechetMere.createObject(u , c);
            if (canClose) {
                c.commit();
            }

        } catch (Exception e) {
            if (canClose && c != null) {
                c.rollback();
            }
            throw e;

        } finally {
            if (canClose && c != null) {
                c.close();
            }
        }
    }
     public void jeter(Connection c, String u,String[] ids) throws Exception {
        boolean canClose = false;
        try {
            if (c == null) {
                c = new UtilDB().GetConn();
                canClose = true;
            }

            for (String id : ids) {
                if (id != null && !id.trim().isEmpty()) {

                    MvtStock mvtstock = (MvtStock) new MvtStock().getById(id, "MVTSTOCK", c);
                    if (mvtstock != null) {
                        mvtstock.jeter(c , u);
                    }
                }
            }

        } catch (Exception e) {
            throw e;
        } finally {
            if (canClose && c != null) {
                c.close();
            }
        }
    }
     

}
